src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -13,33 +13,49 @@ internal sealed class ModuleBuilderImpl : ModuleBuilder
     {
         private readonly Assembly _coreAssembly;
         private readonly string _name;
-        private readonly Dictionary<string, Type> _coreTypes = new();
+        private Type?[] _coreTypes;
         private readonly Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new();
         private readonly Dictionary<Type, TypeReferenceHandle> _typeRefStore = new();
         private readonly List<TypeBuilderImpl> _typeDefStore = new();
         private int _nextMethodDefRowId = 1;
         private int _nextFieldDefRowId = 1;
-
+        private static readonly Type[] s_coreTypes = { typeof(void), typeof(object), typeof(bool), typeof(char), typeof(sbyte), typeof(byte), typeof(short), typeof(ushort), typeof(int),
+                                                        typeof(uint), typeof(long), typeof(ulong), typeof(float), typeof(double), typeof(string), typeof(nint), typeof(nuint) };
         internal ModuleBuilderImpl(string name, Assembly coreAssembly)
         {
             _coreAssembly = coreAssembly;
+            // Use s_coreTypes directly for runtime reflection
+            _coreTypes = (_coreAssembly == typeof(object).Assembly) ? s_coreTypes : new Type[s_coreTypes.Length];
             _name = name;
         }
 
-        internal Type GetTypeFromCoreAssembly(string name)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "Types are preserved via s_coreTypes")]
+        internal Type GetTypeFromCoreAssembly(CoreTypeId typeId)
+        {
+            int index = (int)typeId;
+
+            return _coreTypes[index] ?? (_coreTypes[index] = _coreAssembly.GetType(s_coreTypes[index].FullName!, throwOnError: true)!);
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "Types are preserved via s_coreTypes")]
+        internal int GetTypeIdFromCoreTypes(Type type)
         {
-            Type? type;
+            for(int i=0; i < _coreTypes.Length; i++)
+            {
+                if (_coreTypes[i] == type)
+                {
+                    return i;
+                }
+            }
+
+            Type? foundType = _coreAssembly.GetType(type.FullName!, throwOnError: true);
 
-            // TODO: Use Enum as the key for perf
-            if (!_coreTypes.TryGetValue(name, out type))
+            if (foundType == null || !Enum.TryParse(foundType.Name, out CoreTypeId result))
```

Comment:
It is going to be fairly common to pass non-core types to this method. It means that we will hit this slow enum parsing path frequently. It should be better to populate `_coreTypes` once so that we do not need to keep retrying the parsing for each non-core type.